content/lectures/lec03.md

```diff
@@ -7,4 +7,192 @@ slides: ""
+
+So, if we're unhappy with the Docker CLI as our interface for running containers, let's think about what we want out of a run system:
+
+- Reproducible (should be easy to share)
+- Easy to understand what's going on (human readable)
+- Confidence in making changes to config
+- Resource lifecycle management (we can set parameters that manage our containers)
+
+If we look at the example from the previous section, we can see that apart from resource lifecycle management, we have nothing else from our wishlist.
+
+## Docker Compose
+
+The improvement here is Docker Compose. This is a tool that allows us to "compose" multiple Docker containers together, and manage their lifecycles. Docker Compose manages the lifecycle of multiple containers at the same time, handles all runtime settings for us (environment variables, port maps, volumes, etc), and can be started, stopped and inspected with simple shell commands. Now, if we look back at our wishlist, we meet all of the criteria.
+
+## YAML (YAML Ain't Markup Language)
+
+What does Docker compose actually look like? The configuration is written in a language called YAML. This is great for a lot of reasons: YAML is a simple and human-readable config language, it is standard across many infrastructure tools like Kubernetes and Continuous Integration, and it has types which will prevent us from making some major mistakes. The grammar of YAML is made up of these base types:
+
+- Scalar
+  - String
+  - Number
+  - Boolean
+- Dictionary 
+- List
+
+The top-level structure is always a dictionary, a list of key/value pairs. If you're familiar with JSON, then this should seem familiar. While there are only a few types, by composing these data types we can be quite expressive.
+
+### Scalars
+
+#### Strings
+
+For strings we don't need quotes, however it is generally considered best practice to quote your strings. If you include numbers in your strings and don't quote them, the compiler interprets it as a number, so it's always safer to add quotes.
+```yaml
+armaan: "Tobaccowalla"
+peyton: Walters
+```
+
+#### Multi-line Strings 
+
+There are two operators, `|` (pronounced pipe) and `>` (pronounced chomp). The pipe preserves newlines and whitespace in the YAML, whereas the chomp puts everything on the same line.
+```yaml
+campbell: |
+  This is Campbell's first sentence.
+  This second sentence is on a new line.
+davis: >
+  This is Davis's first sentence.
+  This second sentence will end up
+  being on the same line as the first.
+```
+
+#### Numbers
+
+Fairly straightforward, no quotes and you get a number.
+```yaml
+armaan: 21
+campbell: 20.5
+```
+
+#### Booleans
+
+YAML let's you get away with a lot for boolean values. You should always use `true` or `false`. However, YAML does accept `False` and `NO` but this can make your config confusing and hard to read. Also, these are subject to change in the future so using anything other than `true` or `false` is not particularly maintainable.
```

Comment:
Note for us: I'm pretty sure yaml 1.2 requires booleans to be `true` or `false` but since we can't guarantee that all the tools we work with will only use >=1.2, we should leave this note in